external/fv3fit/fv3fit/keras/_models/shared/halos.py

```diff
@@ -0,0 +1,139 @@
+        return tuple(shape_dict[dim] for dim in dims)
+
+
+def _quantities_from_dataset(
+    ds: xr.Dataset, n_halo: int
+) -> Mapping[Hashable, fv3gfs.util.Quantity]:
+    return_dict = {}
+    nx, ny, nz = ds.dims["x"], ds.dims["y"], ds.dims["z"]
+    sizer = _NoBufferPointGridSizer(
+        nx=nx, ny=ny, nz=nz, n_halo=n_halo, extra_dim_lengths={**ds.sizes}
+    )
+    factory = fv3gfs.util.QuantityFactory(sizer=sizer, numpy=np)
+    for name, da in ds.data_vars.items():
+        quantity = factory.zeros(
+            dims=da.dims, units=da.attrs.get("units", "unknown"), dtype=da.dtype
+        )
+        quantity.view[:] = da.values[:]
+        return_dict[name] = quantity
+    return return_dict
+
+
+def _dataset_from_quantities(
+    quantities: Mapping[Hashable, fv3gfs.util.Quantity]
+) -> xr.Dataset:
+    data_vars = {}
+    for name, quantity in quantities.items():
+        data_vars[name] = xr.DataArray(
+            data=quantity.data, dims=quantity.dims, attrs={"units": quantity.units}
+        )
+    return xr.Dataset(data_vars=data_vars)
+
+
+def _halo_update(
+    communicators: Sequence[fv3gfs.util.CubedSphereCommunicator],
+    quantities: Sequence[fv3gfs.util.Quantity],
+    n_halo: int,
+):
+    req_list = []
+    for comm, quantity in zip(communicators, quantities):
+        req_list.append(comm.start_halo_update(quantity, n_points=n_halo))
+    for req in req_list:
+        req.wait()
+
+
+def append_halos(ds: xr.Dataset, n_halo: int) -> xr.Dataset:
+    """
+    Given a dataset with "tile", "x", and "y" dimensions with no halo data,
+    return a dataset which has n_halo halo points appended to the start and
+    end of the x and y dimensions.
+    """
+    # import pdb;pdb.set_trace()
+    if any(name not in ds.dims for name in ("tile", "x", "y")):
+        raise ValueError(
+            f"dataset must have 'tile', 'x', and 'y' dimensions, got {ds.dims}"
+        )
+    n_tiles = ds.sizes["tile"]
+    if n_tiles != 6:
+        raise ValueError(f"dataset must have 6 tiles to halo update, got {n_tiles}")
+    comms = _create_comms(total_ranks=6)
+    communicators = [
+        fv3gfs.util.CubedSphereCommunicator(
+            comm=comm,
+            partitioner=fv3gfs.util.CubedSpherePartitioner(
+                tile=fv3gfs.util.TilePartitioner(layout=(1, 1))
+            ),
+        )
+        for comm in comms
+    ]
+    datasets_in = [ds.isel(tile=i) for i in range(0, 6)]
+    quantity_dicts = [_quantities_from_dataset(ds, n_halo=n_halo) for ds in datasets_in]
+    for name in ds.data_vars.keys():
+        quantities = [d[name] for d in quantity_dicts]
+        if n_halo > 0:
+            _halo_update(communicators, quantities, n_halo=n_halo)
+    datasets_out = [_dataset_from_quantities(d) for d in quantity_dicts]
+    ds_out = xr.concat(datasets_out, dim="tile")
+    ds_out = match_prediction_to_input_coords(input=ds, prediction=ds_out)
+    # if you don't drop coords here, the model fails to train - no idea why
```

Comment:
One of the grand mysteries of xarray. `xr.testing.assert_identical(ds_out, ds_out.drop(ds_out.coords.keys())` does not raise AssertionError, but if I return the first one (`ds_out`) the model fails to train, staying at around 0.97 loss. Printing `ds_out` shows no coordinates present. Printing `ds_out.coords.keys()` shows a KeysView that says coordinates are empty.